Storage: Support VM image unpack size greater than volume size

tomponline opened the pull request:
This PR allows images larger than the default 10GB block disk size to be used.

Fixes https://github.com/lxc/lxd/issues/7346 

Create a VM image larger than the default 10GB block disk size:
```
lxc init images:ubuntu/focal/cloud v1 --vm
lxc config device override v1 root size=15GB // Enlarge beyond default 10GB block disk size
lxc start v1  // Wait for boot and cloud init to resize rootfs inside guest
lxc exec v1 -- df -h  | grep /dev/sda2
/dev/sda2        14G  796M   13G   6% /
lxc stop v1
lxc publish v1 --alias myvm
lxc delete v1
```

Now check we can create new VM from the image when the storage pool has no `volume.size` set.
```
lxc init myvm v1
sudo zfs get volsize default/images/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896.block
NAME                                                                                   PROPERTY  VALUE    SOURCE
default/images/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896.block  volsize   14.0G    local
sudo zfs get volsize default/virtual-machines/v1.block
NAME                               PROPERTY  VALUE    SOURCE
default/virtual-machines/v1.block  volsize   14.0G    local
```

Now check that this fails if `volume.size` is set on the pool smaller than the image size when the larger optimized snapshot volume already exists.
```
lxc delete v1
lxc storage set default volume.size 9GB
lxc init myvm v1
Creating v1
Error: Create instance from image: Source image size (14999994368) exceeds specified volume size (9000000000)
```

Now check the optimized snapshot volume creation fails for the same reason:

```
lxc storage volume delete default image/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896
lxc init myvm v1
Creating v1
Error: Create instance from image: Source image size (14999994368) exceeds specified volume size (9000000000)
```

Now check we can set pool's `volume.size` larger than image and optimized snapshot and resulting instance volume use the larger size.

```
lxc storage set default volume.size 16GB
lxc init myvm v1
sudo zfs get volsize default/images/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896.block
NAME                                                                                   PROPERTY  VALUE    SOURCE
default/images/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896.block  volsize   14.9G    local
sudo zfs get volsize default/virtual-machines/v1.block
NAME                               PROPERTY  VALUE    SOURCE
default/virtual-machines/v1.block  volsize   14.9G    local
```

Now check we can reduce the pool's `volume.size` back to minimum image size then create new instance, causing old larger optimized volume to be regenerated.

```
lxc delete v1
lxc storage set default volume.size 15GB
lxc init myvm v1
sudo zfs get volsize default/images/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896.block
NAME                                                                                   PROPERTY  VALUE    SOURCE
default/images/f037ad0b451277b594d9f5d80f518f77a6f1a7184a2c7d55939946192ee9a896.block  volsize   14.0G    local
sudo zfs get volsize default/virtual-machines/v1.block
NAME                               PROPERTY  VALUE    SOURCE
default/virtual-machines/v1.block  volsize   14.0G    local
```

tomponline commented:
@stgraber I've hit a bit of a snag with this one. Although this PR does work, it doesn't fix the original issue of allowing a published VM >10GB to be used as the basis of a new VM.

The reason for this is that although this PR allows a new cached optimized image snapshot volume to be greated that is >10GB, when LXD is attempting to use it as the basis of a new VM, it detects that the source snapshot volume is greater than the profile's default size and because it is unable to shrink it, then proceeds to delete it and regenerated it, subsequently failing for the same reason on the 2nd attempt.

I think what we need to do is that for block optimized image volumes, we always create them as the size of the source image (and do not grow them to the profile's default block size), then when we create a VM from those cached images, we will just grow the source volume's snapshot to the desired size (and if the size specified is smaller, then just fail because regenerating isn't going to work for block images).

stgraber commented:
So the problem with that approach is that it makes bulk VM creation significantly slower as we will always need to go through the gpt resize code path rather than only do it for abnormally sized instances.

tomponline commented:
@stgraber OK, how about this:

1. Modify `drivers.VolumeFiller` struct to have a field that can store the `imageVolSize`.
2. As the `VolumeFiller` is passed to the `b.driver.CreateVolume(imgVol, &volFiller, op)` function as a pointer, that function could then update the `imageVolSize` value from `volFiller.Fill()` by changing the definition of `volFiller.Fill()` to return the imageVolSize it discovered whilst unpacking the image.
3. When `b.driver.CreateVolume(imgVol, &volFiller, op)` is called from `EnsureImage()` we can use the `imageVolSize` to store in the image's volume DB record as config, such as `image_vol_size`.
4. This config option can be inspected in `CreateInstanceFromImage` during the optimized volume creation, after `EnsureImage` is called, by loading the created volume DB record. We can then set this as the `size` property on the `imgVol` passed to `b.driver.CreateVolumeFromCopy(vol, imgVol, false, op)`.
5. Inside the drivers, we can then decide based on the requested volume size (or profile's default volume size) whether it is below the imgVol size and if so, we can override the size of the volume created.
6. For the non-optimized volume scenario, we can do similar, but can directly use the `imageVolSize` from the result of the filler inside the driver to decide whether to use it to override the volume's `size` setting before creating it.

tomponline commented:
jenkins: test this please

lxc-jenkins commented:
Testsuite passed

tomponline commented:
@stgraber please can you take a look at this commit https://github.com/lxc/lxd/pull/7380/commits/b7118bc787bbbcbff584e015bbedc57cbfbb7bd1, it is my implementation of my interpretation of your post here https://github.com/lxc/lxd/issues/7346#issuecomment-631061826

If this is OK I will then use that for checking/deriving the size of new volumes in the drivers (not just LVM).

stgraber commented:
Logic looks good to me (in that commit).

lxc-jenkins commented:
Testsuite passed

tomponline commented:
@stgraber this is ready for review now, I've added some logic unit tests for `Volume.ConfigSizeFromSource` to document my understanding of that function. And I've updated the PR description with my test plan examples.
